ArrayPixelBuffer superclass of TypedArrayPixelBuffer

squelart opened the pull request:
#### 921a80bb8d6adfbad24f3707282965b8b4080274
<pre>
ArrayPixelBuffer superclass of TypedArrayPixelBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=296971">https://bugs.webkit.org/show_bug.cgi?id=296971</a>
<a href="https://rdar.apple.com/157605644">rdar://157605644</a>

Reviewed by Mike Wyrzykowski.

Currently PixelBuffer has 3 direct subclasses: ByteArrayPixelBuffer,
Float16ArrayPixelBuffer and ShareablePixelBuffer.

Byte- and Float16- have pretty much the same underlying storage: A
JavaScriptCore typed array, which is always a subclass of
JSC::ArrayBufferView. (But ShareablePixelBuffer is different.)

In some situations that only deal, or used to deal, with
ByteArrayPixelBuffer, we will want them to work with
Float16ArrayPixelBuffer as well.
Instead of doubling the code at each site, it would be best to use a
common base class that can abstract the differences in most cases, like
accessing the data as bytes.
But at the same time, we can&apos;t just use the current base PixelBuffer, as
we don&apos;t want to have to also deal with ShareablePixelBuffer, which
shouldn&apos;t happen in these code paths anyway.

So this introduces an intermediate abstract base class ArrayPixelBuffer
for these array-based concrete PixelBuffers.
Note that this abstract class ArrayPixelBuffer has non-virtual member
functions to access the data, so there is no penalty when calling them,
and the compiler could even inline them.

In summary, the class hierarchy is now as follows:
- PixelBuffer: Any kind of pixel buffer, this abstract base class
  stores the format, size, and a non-owning span to the data bytes.
- ShareablePixelBuffer: Concrete subclass owning a SharedMemory,
  with non-virtual data() to access that SharedMemory.
- ArrayPixelBuffer: Abstract subclass owning a JSC::ArrayBufferView,
  with non-virtual data() to access that JSC buffer. Useful for e.g.:
  canvas to cache putImageData pixels.
~ TypedArrayPixelBuffer: Templated concrete subclass, which
  guarantees by construction that the underlying data() is of a
  certain concrete type.
- ByteArrayPixelBuffer=Typed&lt;JSC::Uint8ClampedArray&gt;:
  Instantiation for byte-oriented buffers, useful for code that
  only deals with these, like BitmapImage.
- Float16ArrayPixelBuffer=Typed&lt;JSC::Float16Array&gt;: The newer
  intantiation to be used for HDR-capable elements.

Other improvements:
- Make leaf subclasses final.
- Make PixelBuffer member variables private for better encapsulation.
- Make PixelBuffer member variables const because they should never change.
- Make PixelBuffer::type() pure virtual, to ensure concrete subclasses
  override it.
- Qualify takeData() as `&amp;&amp;` to force object to be moved-from and catch
  potential use-after-move&apos;s.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageDataArray.cpp:
(isType): Deleted.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
* Source/WebCore/platform/graphics/ArrayPixelBuffer.cpp: Added.
(WebCore::ArrayPixelBuffer::ArrayPixelBuffer):
* Source/WebCore/platform/graphics/ArrayPixelBuffer.h: Added.
(WebCore::ArrayPixelBuffer::takeData):
(isType):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::type const): Deleted.
* Source/WebCore/platform/graphics/TypedArrayPixelBuffer.cpp:
(WebCore::&gt;::TypedArrayPixelBuffer):
* Source/WebCore/platform/graphics/TypedArrayPixelBuffer.h:
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.h:
(WebKit::ShareablePixelBuffer::data const): Deleted.

Canonical link: <a href="https://commits.webkit.org/317522@main">https://commits.webkit.org/317522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2ae3734773080c3f0f66b02bea4ad4b2bb8ff06

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/175468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/49400 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/43181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/185054 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/137614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a8ebd1bb-1b45-465e-a339-7b8809f1253a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/177332 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/50692 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/49083 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/136618 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/137614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/330a6dd3-3abe-4df1-b863-6b3cc7dbe0b7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/178425 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/40035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/157373 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/117096 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/11ef5a68-8a68-4b48-b423-d05d32ffee7a) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/38677 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/37060 "Passed tests") | [⏳ 🧪 jsc-wpe](https://ews-build.webkit.org/#/builders/JSC-Tests-WPE-EWS "Waiting to run tests") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/149099 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/36866 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/33529 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/41087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/41263 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/187479 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/49067 "Built successfully") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/145585 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/39184 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/49747 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/33490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/145423 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/40239 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/121940 "Built successfully") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/115667 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/48253 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/11842 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/47656 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/47886 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/47713 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f431ae438c9791cc8260a7dfd4f3955d4814e74b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/504f5df0e6018e67dec6fd7e6ff37dd2415ca569. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d778c5d952a92da656a86ab6433d72dff723f2c3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

squelart committed:
Templatize Byte/Float16ArrayPixelBuffer, subclasses of ArrayPixelBuffer
https://bugs.webkit.org/show_bug.cgi?id=296971
rdar://157605644

Reviewed by NOBODY (OOPS!).

Currently PixelBuffer has 3 direct subclasses: ByteArrayPixelBuffer,
Float16ArrayPixelBuffer and ShareablePixelBuffer.

For a start: Byte- and Float16- have pretty much the same underlying
storage: A JavaScriptCore typed array, always a subclass of
JSC::ArrayBufferView. (But ShareablePixelBuffer is different.)

In some situations that only deal (or used to deal) with
ByteArrayPixelBuffer, we now want them to work with
Float16ArrayPixelBuffer as well.
Instead of doubling the code at each site, it would be best to use a
common base class that can abstract the differences in most cases, like
accessing the data as bytes.
But at the same time, we can't just use the current base PixelBuffer, as
we don't want to have to also deal with ShareablePixelBuffer, which
shouldn't happen in these code paths anyway.

So this introduces an intermediate abstract base class ArrayPixelBuffer
for these array-based concrete PixelBuffers -- Byte and Float16 for now,
but in the future Float32 could be added more easily.
Note that this abstract class ArrayPixelBuffer has non-virtual member
functions to access the data, so there is no penalty when calling them,
and the compiler could even inline them.

Its first use is to cache ImageData in 2D canvas, so it will be possible
to transparently cache either uint8 or float16 data.

In addition, this patch refactors the existing Byte- and
Float16ArrayPixelBuffer's into a common template class of which they're
instantiations.

In summary, the class hierarchy is now as follows:
- PixelBuffer -- Any kind of pixel buffer, this abstract base class
  stores the format, size, and a non-owning span to the data bytes.
  - ShareablePixelBuffer -- Concrete subclass owning a SharedMemory,
    with non-virtual data() to access that SharedMemory.
  - ArrayPixelBuffer -- Abstract subclass owning a JSC::ArrayBufferView,
    with non-virtual data() to access that JSC buffer. Useful for e.g.:
    canvas to cache putImageData pixels.
    ~ TypedArrayPixelBuffer -- Templated concrete subclass, which
      guarantees by construction that the underlying data() is of a
      certain concrete type.
    - ByteArrayPixelBuffer=Typed<JSC::Uint8ClampedArray>
      -- Instantiation for byte-oriented buffers, useful for code that
      only deals with these, like BitmapImage.
    - Float16ArrayPixelBuffer=Typed<JSC::Float16Array> -- The newer
      intantiation to be used for HDR-capable elements.

Other improvements:
- Make leaf subclasses final.
- Make PixelBuffer member variables private for better encapsulation.
- Make PixelBuffer member variables const because they should never change.
- Make PixelBuffer::type() pure virtual, to ensure subclasses override it.
- Provide PixelBuffer::bytesPerPixelComponent(PixelFormat), and start
  using it in some size checks that assumed 1 byte per component.
- Make ImageDataArray (and therefore ImageData) constructible from a
  JSC::ArrayBufferView.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageDataArray.cpp:
(isType): Deleted.
* Source/WebCore/html/ImageDataArray.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::CachedContentsImageData::CachedContentsImageData):
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/ArrayPixelBuffer.cpp: Added.
(WebCore::ArrayPixelBuffer::ArrayPixelBuffer):
(WebCore::ArrayPixelBuffer::create):
* Source/WebCore/platform/graphics/ArrayPixelBuffer.h: Added.
(WebCore::ArrayPixelBuffer::protectedData const):
(WebCore::ArrayPixelBuffer::takeData):
(isType):
* Source/WebCore/platform/graphics/ByteArrayPixelBuffer.cpp: Removed.
* Source/WebCore/platform/graphics/ByteArrayPixelBuffer.h:
(): Deleted.
(WebCore::ByteArrayPixelBuffer::protectedData const): Deleted.
(WebCore::ByteArrayPixelBuffer::takeData): Deleted.
* Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.cpp: Removed.
* Source/WebCore/platform/graphics/Float16ArrayPixelBuffer.h:
(): Deleted.
(WebCore::Float16ArrayPixelBuffer::takeData): Deleted.
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::computeRawPixelComponentCount):
(WebCore::PixelBuffer::computeBufferSize):
(WebCore::PixelBuffer::PixelBuffer):
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::bytesPerPixelComponent):
(WebCore::PixelBuffer::componentsPerPixel):
(WebCore::PixelBuffer::bytesPerPixel):
(WebCore::PixelBuffer::type const): Deleted.
* Source/WebCore/platform/graphics/TypedArrayPixelBuffer.h: Added.
(WebCore::>::create):
(WebCore::>::tryCreate):
(WebCore::>::TypedArrayPixelBuffer):
(WebCore::>::createScratchPixelBuffer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToVideoElement):
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.h: Missing include, due to added files in this patch.
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.cpp:
(WebKit::ShareablePixelBuffer::ShareablePixelBuffer):
(WebKit::ShareablePixelBuffer::createScratchPixelBuffer const):
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.h:
(WebKit::ShareablePixelBuffer::data const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d704ede19b283fd589d3eb0a590cd70215388dd6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9a709e8d51c340469f19e605e736f3b62078522a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd50cfaf41d0cf3d77ba2fba202b3266765d5c46. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2187c55a3df7a98d913d04f7bfbb7a046d11429f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48950

squelart commented:
@squelart note to self: When coming back to this PR here, read review comments at https://github.com/WebKit/WebKit/pull/49050#pullrequestreview-3106791470 and integrate them here if still relevant.

squelart committed:
ArrayPixelBuffer superclass of TypedArrayPixelBuffer
https://bugs.webkit.org/show_bug.cgi?id=296971
rdar://157605644

Reviewed by Mike Wyrzykowski.

Currently PixelBuffer has 3 direct subclasses: ByteArrayPixelBuffer,
Float16ArrayPixelBuffer and ShareablePixelBuffer.

Byte- and Float16- have pretty much the same underlying storage: A
JavaScriptCore typed array, which is always a subclass of
JSC::ArrayBufferView. (But ShareablePixelBuffer is different.)

In some situations that only deal, or used to deal, with
ByteArrayPixelBuffer, we will want them to work with
Float16ArrayPixelBuffer as well.
Instead of doubling the code at each site, it would be best to use a
common base class that can abstract the differences in most cases, like
accessing the data as bytes.
But at the same time, we can't just use the current base PixelBuffer, as
we don't want to have to also deal with ShareablePixelBuffer, which
shouldn't happen in these code paths anyway.

So this introduces an intermediate abstract base class ArrayPixelBuffer
for these array-based concrete PixelBuffers.
Note that this abstract class ArrayPixelBuffer has non-virtual member
functions to access the data, so there is no penalty when calling them,
and the compiler could even inline them.

In summary, the class hierarchy is now as follows:
- PixelBuffer: Any kind of pixel buffer, this abstract base class
  stores the format, size, and a non-owning span to the data bytes.
- ShareablePixelBuffer: Concrete subclass owning a SharedMemory,
  with non-virtual data() to access that SharedMemory.
- ArrayPixelBuffer: Abstract subclass owning a JSC::ArrayBufferView,
  with non-virtual data() to access that JSC buffer. Useful for e.g.:
  canvas to cache putImageData pixels.
~ TypedArrayPixelBuffer: Templated concrete subclass, which
  guarantees by construction that the underlying data() is of a
  certain concrete type.
- ByteArrayPixelBuffer=Typed<JSC::Uint8ClampedArray>:
  Instantiation for byte-oriented buffers, useful for code that
  only deals with these, like BitmapImage.
- Float16ArrayPixelBuffer=Typed<JSC::Float16Array>: The newer
  intantiation to be used for HDR-capable elements.

Other improvements:
- Make leaf subclasses final.
- Make PixelBuffer member variables private for better encapsulation.
- Make PixelBuffer member variables const because they should never change.
- Make PixelBuffer::type() pure virtual, to ensure concrete subclasses
  override it.
- Qualify takeData() as `&&` to force object to be moved-from and catch
  potential use-after-move's.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageDataArray.cpp:
(isType): Deleted.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
* Source/WebCore/platform/graphics/ArrayPixelBuffer.cpp: Added.
(WebCore::ArrayPixelBuffer::ArrayPixelBuffer):
* Source/WebCore/platform/graphics/ArrayPixelBuffer.h: Added.
(WebCore::ArrayPixelBuffer::takeData):
(isType):
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
* Source/WebCore/platform/graphics/PixelBuffer.h:
(WebCore::PixelBuffer::type const): Deleted.
* Source/WebCore/platform/graphics/TypedArrayPixelBuffer.cpp:
(WebCore::>::TypedArrayPixelBuffer):
* Source/WebCore/platform/graphics/TypedArrayPixelBuffer.h:
* Source/WebKit/GPUProcess/graphics/ShareablePixelBuffer.h:
(WebKit::ShareablePixelBuffer::data const): Deleted.

Canonical link: https://commits.webkit.org/317522@main
